[c10d] Remove init_process_group from DeviceMesh

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #162571
* __->__ #162549
* #163058
* #162770
* #162545
* #162529

As titled.
So that `DeviceMesh` creation will no longer implicitly create a global process group, which is time consuming and most-of-time not used.
`init_process_group` is replaced with `dist.init()` -- just for `get_rank()` and `get_world_size()` to work.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

kwen2501 committed:
Update

[ghstack-poisoned]

fduwjj commented:
lol can you fix CI failure?

Skylion007 commented on torch/distributed/device_mesh.py:
```diff
@@ -447,15 +447,34 @@ def __init__(
+                    "Distributed environment is not initialized, initializing "
+                    "it during DeviceMesh creation with an assumption that "
+                    "the DeviceMesh has the same scope as the world."
+                )
```
Pass exc_info=True for better logging

kwen2501 committed:
Update on "[c10d] Remove init_process_group from DeviceMesh"


As titled.
So that `DeviceMesh` creation will no longer implicitly create a global process group, which is time consuming and most-of-time not used.
`init_process_group` is replaced with `dist.init()` -- just for `get_rank()` and `get_world_size()` to work.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]
